File: src/leetcode/ClosestBinarySearchTreeValue.java
package leetcode;

import org.junit.Test;

import static junit.framework.TestCase.assertEquals;


public class ClosestBinarySearchTreeValue {

    public int closestValue(TreeNode root, double target) {
        TreeNode node = root;
        double diff = Math.abs(root.val - target);
        while(true) {
            if (node.val > target) {
                if (node.left == null) return node.val;
                double leftDiff = Math.abs(node.left.val - target);
                if (leftDiff > diff) return node.val;
                else {
                    node = node.left;
                    diff = leftDiff;
                }
            } else {
                if (node.right == null) return node.val;
                double rightDiff = Math.abs(node.right.val - target);
                if (rightDiff > diff) return node.val;
                else {
                    node = node.right;
                    diff = rightDiff;
                }
            }
        }
    }

    @Test
    public void test() {
        TreeNode root = new TreeNode(13);
        root.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.left.left = new TreeNode(1);
        root.left.left.right = new TreeNode(2);
        root.left.left.right.right = new TreeNode(3);
        root.left.left.left = new TreeNode(0);

        assertEquals(3, closestValue(root, 3.14));

    }
}

File: src/leetcode/ClosestBinarySearchTreeValue.kt
package leetcode

import junit.framework.TestCase
import org.junit.Test
import kotlin.math.abs

internal class ClosestBinarySearchTreeValue {
    fun closestValue2(root: TreeNode, target: Double): Int {
        return s(root, target, makePair(root.`val`, target)).first
    }

    private fun s(node: TreeNode, target: Double, cur: Pair<Int, Double>): Pair<Int, Double> {
        node.left?.apply {
            if (node.`val` > target)
                s(this, target, makePair(this.`val`, target)).apply {
                    if (this.second < cur.second) return this
                }
        }
        node.right?.apply {
            if (node.`val` < target) {
                s(this, target, makePair(this.`val`, target)).apply {
                    if (this.second < cur.second) return this
                }
            }
        }
        return cur
    }

    private fun makePair(value: Int, target: Double): Pair<Int, Double> {
        return Pair(value, abs(target - value))
    }

    fun closestValue(root: TreeNode, target: Double): Int {
        val a = root.`val`
        val kid = (if (target < a) root.left else root.right) ?: return a
        val b = closestValue(kid, target)
        return if (abs(a - target) < abs(b - target)) a else b
    }

    @Test
    fun test() {
        val root = TreeNode(13)
        root.left = TreeNode(4)
        root.left.right = TreeNode(5)
        root.left.left = TreeNode(1)
        root.left.left.right = TreeNode(2)
        root.left.left.right.right = TreeNode(3)
        root.left.left.left = TreeNode(0)
        TestCase.assertEquals(3, closestValue2(root, 3.14))
    }

    @Test
    fun test22() {

    }
}